core/storage/src/trie.rs

```diff
@@ -246,63 +314,149 @@ impl Trie {
         }
     }
 
-    fn print_node(&self, node_storage: &TrieNodeStorage, node: &TrieNode) {
-        println!("{:?}", node);
-        match node {
-            TrieNode::Branch(branch) => {
+    fn delete(&self, node_storage: &mut TrieNodeStorage, node: &TrieNode, nibbles: &[u8], position: usize) -> Result<Option<CryptoHash>, String> {
+        match node.data {
+            TrieNodeData::Empty => {
+                Err("Removing empty node".to_string())
+            },
+            TrieNodeData::Leaf(ref leaf) => {
+                if *leaf.key == nibbles[position..] {
+                    node_storage.delete(node);
+                    Ok(None)
+                } else {
+                    Err("Deleting missing leaf node".to_string())
+                }
+            },
+            TrieNodeData::Branch(ref branch) => {
+                if position == nibbles.len() {
+                    node_storage.delete(node);
+                    let mut branch_node = TrieBranch { children: [None; 16], value: None };
+                    branch_node.children.clone_from_slice(&branch.children);
+                    if branch_node.children.iter().filter(|x| x.is_some()).count() == 0 {
+                        Ok(None)
+                    } else {
+                        Ok(Some(node_storage.add(TrieNodeData::Branch(branch_node))))
+                    }
+                } else {
+                    let hash = branch.children[nibbles[position] as usize];
+                    match hash {
```

Comment:
It is probably better to match on something like `hash.and_then(|h| node_storage.get(&h)).map(|sub_node| self.delete(node_storage, &sub_node, nibbles, position + 1))`